src/common/datomish/transact.cljc

```diff
@@ -552,12 +567,75 @@
     (:db-after (<? (<with db tx-data)))))
 
 (defn <transact!
+  "Submits a transaction to the database for writing.
+
+  Returns a pair-chan resolving to `[result error]`."
   [conn tx-data]
   {:pre [(conn? conn)]}
-  (let [db (db conn)] ;; TODO: be careful with swapping atoms.
-    (db/in-transaction!
-      db
-      #(go-pair
-         (let [report (<? (<with db tx-data))]
-           (reset! (:current-db conn) (:db-after report))
-           report)))))
+  (let [result (a/chan 1)]
+    ;; Any race to put! is a real race between callers of <transact!.  We can't just park on put!,
+    ;; because the parked putter that is woken is non-deterministic.
+    (a/put! (:transact-chan conn) [tx-data result])
+    result))
+
+(defn- start-transactor [conn]
+  (let [token-chan (a/chan 1)]
+    (go
+      (>! token-chan (gensym "transactor-token"))
+      (loop []
+        (let [token (<! token-chan)]
+          (when-let [[tx-data result] (<! (:transact-chan conn))]
+            (let [pair
+                  (<! (go-pair ;; Catch exceptions, return the pair.
+                        (let [db (db conn)
+                              report (<? (db/in-transaction!
+                                           db
+                                           #(-> (<with db tx-data))))]
+                          ;; We only get here if the transaction is committed.
+                          (reset! (:current-db conn) (:db-after report))
+                          (>! (:listener-source conn) report)
+                          report)))]
+              (>! result pair))
+            (a/close! result)
```

Comment:
Closing a closed channel is a no-op.  I suppose we could do this in a `try` block, but let's see that be a problem in the wild first.